### ISSUE TITLE
1.8.1 deletes interface addrs from tailscale0 with NetworkManager DNS management (Debian Buster, Ubuntu 20.04)

### ISSUE THREAD

**Issue by mryellow** [tailscale/tailscale#1875]
"DHCP Failed to Start"

**To Reproduce**

1. Update via Debian repos

**Version information:**
 - Device: Laptop
 - OS: Debian Buster
 - Tailscale version:

```
1.8.1
  tailscale commit: fbfdcc56edb92066ef6dbaa1ae9d9bffa5455757
  other commit: c60d024101bc5ca01ddfe31194aaa6a2c02c7dd1
  go version: go1.16.3-ts2c6ff2c
```

`BUG-77dd5c092cb927c7f69001df2692cfdce3d348e92ecb099ccc85a13120f514f0-20210507005355Z-31b47c7cbe6190cf`

**Comment by mryellow** [comment]
Repo in question: `deb https://pkgs.tailscale.com/stable/debian buster main`

**Comment by mryellow** [comment]
```
cat /etc/debian_version 
10.9
```

> Could you provide a little more detail of what broke? 

Don't know. It's not working anymore, downgrade and it works again.

**Comment by mryellow** [comment]
```
sudo tailscale down
sudo tailscale up
```

Spits out:

`context canceled`

...

Here's something. `failure to start DHCP: missing MAC address`. Replaced potentially PII with `...GUID...`.

`sudo apt-get install tailscale=1.8.1`
`/var/log/user.log`

```
May  7 11:24:52 debian NetworkManager[1069]: <info>  [1620350692.2821] device (tailscale0): state change: activated -> unmanaged (reason 'connection-assumed', sys-iface-state: 'managed')
May  7 11:24:57 debian NetworkManager[1069]: <info>  [1620350697.7038] manager: (tailscale0): new Tun device (/org/freedesktop/NetworkManager/Devices/9)
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6610] device (tailscale0): state change: unmanaged -> unavailable (reason 'connection-assumed', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6684] keyfile: add connection /run/NetworkManager/system-connections/tailscale0.nmconnection (...GUID...,"tailscale0")
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6706] device (tailscale0): state change: unavailable -> disconnected (reason 'connection-assumed', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6729] device (tailscale0): Activation: starting connection 'tailscale0' (...GUID...)
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6735] device (tailscale0): state change: disconnected -> prepare (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6759] device (tailscale0): state change: prepare -> config (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6928] device (tailscale0): state change: config -> ip-config (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6931] device (tailscale0): state change: ip-config -> ip-check (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6969] device (tailscale0): state change: ip-check -> secondaries (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.6972] device (tailscale0): state change: secondaries -> activated (reason 'none', sys-iface-state: 'external')
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.8028] device (tailscale0): Activation: successful, device activated.
May  7 11:24:59 debian NetworkManager[1069]: <warn>  [1620350699.8178] device (tailscale0): failure to start DHCP: missing MAC address
May  7 11:24:59 debian NetworkManager[1069]: <info>  [1620350699.8178] device (tailscale0): state change: activated -> failed (reason 'dhcp-start-failed', sys-iface-state: 'managed')
May  7 11:25:00 debian NetworkManager[1069]: <warn>  [1620350700.0301] device (tailscale0): Activation: failed for connection 'tailscale0'
May  7 11:25:00 debian NetworkManager[1069]: <warn>  [1620350700.0323] device (tailscale0): Failed to apply IPv4 configuration
May  7 11:25:00 debian NetworkManager[1069]: <info>  [1620350700.0368] audit: op="device-reapply" interface="tailscale0" ifindex=9 args="ipv6.method,ipv6.dns-priority,ipv6.never-default,ipv6.addresses,ipv6.ignore-auto-dns,ipv6.ignore-auto-routes,ipv6.dns-search,ipv4.method,ipv4.dns-priority,ipv4.never-default,ipv4.ignore-auto-dns,ipv4.ignore-auto-routes,ipv4.dns-search" pid=12708 uid=0 result="success"
May  7 11:25:00 debian NetworkManager[1069]: <info>  [1620350700.2838] device (tailscale0): state change: failed -> disconnected (reason 'none', sys-iface-state: 'managed')
May  7 11:25:00 debian NetworkManager[1069]: <info>  [1620350700.2938] device (tailscale0): state change: disconnected -> unmanaged (reason 'connection-assumed', sys-iface-state: 'managed')
```

**Comment by bradfitz** [comment]
@danderson,

```
2021-05-07 11:24:57.7013019 +1000 +1000: Program starting: v1.8.1-tfbfdcc56e-gc60d02410, Go 1.16.3-ts2c6ff2c: []string{"/usr/sbin/tailscaled", "--state=/var/lib/tailscale/tailscaled.state", "--socket=/run/tailscale/tailscaled.sock", "--port", "41641"}
2021-05-07 11:24:57.701332207 +1000 +1000: LogID: 77dd5c092cb927c7f69001df2692cfdce3d348e92ecb099ccc85a13120f514f0
2021-05-07 11:24:57.701350363 +1000 +1000: logpolicy: using $STATE_DIRECTORY, "/var/lib/tailscale"
2021-05-07 11:24:57.702493883 +1000 +1000: wgengine.NewUserspaceEngine(tun "tailscale0") ...
2021-05-07 11:24:57.712167747 +1000 +1000: router: v6nat = true
2021-05-07 11:24:57.719578354 +1000 +1000: dns: [rc=nm nm=yes nm-old=yes ret=*dns.nmManager]
2021-05-07 11:24:57.719764271 +1000 +1000: dns: using *dns.nmManager
2021-05-07 11:24:57.719845362 +1000 +1000: link state: interfaces.State{defaultRoute=wlp3s0 ifs={docker0:[172.17.0.1/16] wlp3s0:[192.168.1.101/24]} v4=true v6global=false}
```

Can you take a look? (see BUG marker above)

**Comment by danderson** [comment]
Bah, another set of mutually-exclusive NM configs.

At this point I'm tempted to just remove attempts to configure things with NetworkManager, but that's not going to work well either...

**Comment by danderson** [comment]
Okay, the actual error is a lot of these while the network engine is running:

```
Failed to write packet to TUN device: write /dev/net/tun: input/output error
```

So, the tun device is unhappy. I suspect because NM tore down the interface when it failed to DHCP, which is not its behavior on either ubuntu 18.04 or 21.04, delightfully.

I'm building a 20.04 VM now to reproduce and figure out what sequence of operations make NM happy there, and how to distinguish it and 18.04's behavior. For now I've removed 1.8.1 from the stable track, so that more folks don't run into this while I figure out a fix.

**Comment by danderson** [comment]
Right, as suspected, 7d16c8228bcf70a3b82afe994e8c5fa42057c5f3 is the culprit. On Ubuntu 18.04, we must set an IPv4 management mode in order to be allowed to provide DNS settings, and the DHCP configuration fails gracefully (i.e. fails, but in a way that lets the interface continue to exist, along with its DNS settings).

On a more recent version of NM (my 20.04 VM has 1.22.10), this behavior changed: you no longer have to set an IP management mode to provide DNS settings, but if you _do_ provide an IP management mode, and configuration fails (e.g. DHCP can't run), then NM nuked all address configuration on tailscale0, which obviously breaks everything.

So, we have a setup where DNS configuration must specify an IP management mode on one version of NM, and absolutely must not on another, and getting it wrong results in breaking connectivity. Bah.

Next step is to read NetworkManager's changelogs and see if I can find which version changed this behavior, so that we can add a version gate and provide it with the right configuration. Alternatively, we can try setting DNS config without setting an IP management mode, and parse the returned D-Bus error to see if we need to provide a mode.

**Comment by danderson** [comment]
on 18.04, NetworkManager is version 1.10.6. So, somewhere between 1.10.6 and 1.22.10 is what I'm looking for. Pretty rubbish range :/ I might have to bisect non-LTS releases until I find the less-broken one.

**Comment by danderson** [comment]
@cg31 @mryellow could you try the unstable build at https://pkgs.tailscale.com/unstable and see if it fixes your issues? I think that revert is right, and it seems to work on my zoo of test VMs, but I'd love some confirmation before I release it as 1.8.2.

**Comment by cg31** [comment]
@danderson, it works. 
# tailscale version
1.9.17
  tailscale commit: 3c8e230ee1ea856742647d5e1e79087f100eb01b
  other commit: 2308d3cbc6281d2d4733732d9689a5a518080b74
  go version: go1.16.3-ts2c6ff2c

But there is an odd thing: on a machine with NetworkManager, after I run "tailscale down", then I start a new ssh connecting to the machine (from a physicsl IP), the tailscale0 interface still shows up, although "tailscale status" shows it is down...

**Comment by danderson** [comment]
The tailscale0 interface sticks around even when our network engine is down, just so we don't have to do a bunch of dancing to delete and recreate it. But it should have no IP addresses, and tailscaled has no routing state (so even if the OS sends us stuff, we'll just drop it on the floor).

Thanks for the confirmation that 1.9.17 fixes things. I'll tentatively plan on releasing 1.8.2 tomorrow with this, unless I find another regression.

**Comment by spopinski** [comment]
Confirming that updatng to the latest unstable works on 20.04

**Comment by gnanderson** [comment]
Thanks for your efforts on this @danderson 

with 20.04,  nm 1.22.10, tailscale 1.9.17 I no longer see any issues in the logs with tun device

FYI (no problems, just fyi) `systemctl restart tailscaled` BUG-2170dfd37f5ca805e121cf9d4b8f5a0fd999331e12c6a8eccf85dcfe61f8688e-20210507231246Z-0bde40e9d8185314

**Comment by danderson** [comment]
Released 1.8.2 with this fix (among others). Thanks for bearing with us!

**Comment by mryellow** [comment]
Sorry didn't have time to test earler. Looks to be working.

I note a series of these spammed in user log.

43 of them within a second after `Activation: successful, device activated.`

```
May 11 09:03:18 debian NetworkManager[1069]: <info>  [1620687798.8734] audit: op="device-reapply" interface="tailscale0" ifindex=11 pid=15450 uid=0 result="fail" reason="The settings specified are invalid: ipv4.dns-search: this property is not allowed for 'method=disabled'"
```
